corehq/apps/data_interfaces/static/data_interfaces/js/case_property_input.js

```diff
@@ -0,0 +1,78 @@
+/*
+ * Component to input for a case property.
+ *
+ * If the data dictionary is turned on, this input will be a select2
+ * populated with case properties from the data dictionary. It will not
+ * include system properties such as case name or date opened.
+ * If the data dictionary is not turned on, the input will be a text input.
+ *
+ * Required parameters
+ * - caseTypeObservable: observable storing the relevant case type, necessary so
+ *   that autocomplete options can be updated when the case type changes
+ * - valueObservable: observable to use as the input's value
+ *
+ * Usage
+ * - The page must have an "all_case_properties" item provided to initial page data,
+ *   which should be an object where keys are case type names and values are arrays
+ *   of case property names.
+ * - The calling code must call the `register` function to initialize the component
+ *   Note this must be done on DOM ready, so taht initial page data is available.
+ * - The widget itself is a custom HTML element:
+ *     <case-property-input params="
+ *       valueObservable: name,
+ *       caseTypeObservable: $root.caseType,
+ *     "></case-property-input>
+ */
+hqDefine('data_interfaces/js/case_property_input', [
+    'jquery',
+    'knockout',
+    'underscore',
+    'hqwebapp/js/initial_page_data',
+    'hqwebapp/js/toggles',
+    'hqwebapp/js/select2_knockout_bindings.ko',
+], function (
+    $,
+    ko,
+    _,
+    initialPageData,
+    toggles
+) {
+    var component = {
+        viewModel: function (params) {
+            var self = this;
+
+            self.caseTypeObservable = params.caseTypeObservable;
+            self.valueObservable = params.valueObservable;
+
+            self.allCaseProperties = initialPageData.get("all_case_properties");
```

Comment:
Thoughts on making this strict, to make sure implementers remember to supply it?
```suggestion
            self.allCaseProperties = initialPageData.get("all_case_properties", true);
```

Reply:
I forgot about strict initial page data. But I think that similar to [this](https://github.com/dimagi/commcare-hq/pull/31854#discussion_r912212573), I'd rather the failure mode be "no autocomplete" than a js error.